R/get_predicted_args.R

```diff
@@ -0,0 +1,217 @@
+      '`get_predicted(model, predict = NULL, type = "response")`'
+    ))
+  }
+
+  # copy "type" to "predict"
+  if (!is.null(dots$type)) {
+    predict <- dots$type
+  }
+
+  if (length(predict) > 1) {
+    predict <- predict[1]
+    if (isTRUE(verbose)) {
+      msg <- format_message(sprintf("More than one option provided in `predict`. Using first option '%s' now."), predict[1])
+      warning(msg, call. = FALSE)
+    }
+  }
+
+  # Get info
+  info <- model_info(x, verbose = FALSE)
+
+  # Data
+  if (!is.null(newdata) && is.null(data)) data <- newdata
+  if (is.null(data)) data <- get_data(x, verbose = verbose)
+
+  # CI
+  if (is.null(ci)) ci <- 0
+
+  # check `predict` user-input
+  predict_method <- lapply(
+    class(x), function(i) {
+      tryCatch(utils::getS3method("predict", i),
+               error = function(e) NULL)
+    }
+  )
+  predict_method <- tryCatch(predict_method[!sapply(predict_method, is.null)][[1]], error = function(e) NULL)
+
+  # check aliases
+  if (predict %in% c("expected", "response")) {
+    predict <- "expectation"
+  }
+  if (predict == "predicted") {
+    predict <- "prediction"
+  }
+
+  ## TODO remove in a later update
+  # backward compatibility
+  if (identical(predict, "relation")) {
+    message(format_message(
+      '`predict = "relation" is deprecated.',
+      'Please use `predict = "expectation" instead.'
+    ))
+    predict <- "expectation"
+  }
+
+  # Warn if get_predicted() is not called with an easystats- or
+  # model-supported predicted type
+  easystats_methods <- c("expectation", "link", "prediction", "classification")
+  type_methods <- suppressWarnings(eval(formals(predict_method)$type))
+  supported <- c(easystats_methods, type_methods)
+  if (isTRUE(verbose) && !is.null(predict) && !predict %in% supported) {
+    msg <- format_message(
+      sprintf('`predict` = "%s"` is not officially supported by `get_predicted()`.', predict),
+      '`predict` will be passed directly to the `predict()` method for the model and not validated.',
+      'Please check the validity and scale of the results.',
+      'Set `verbose = FALSE` to silence this warning, or use one of the supported values for the `predict` argument:',
+      paste(" ", paste(sprintf('"%s"', setdiff(easystats_methods, c("expected", "predicted"))), collapse = ", "))
+    )
+    warning(msg, call. = FALSE)
+  }
+
+
+  transform <- FALSE
+
+  # type_arg  = what we pass own to type
+  # scale_arg = the scale of the predictions (link, response, terms, ...)
+  #             this may differ from type_arg, because we back-transform
+  # transform = whether we back-transform predictions, SE and CI
+
+  # define prediction type (that's for the initial call to stats::predict)
+
+  # terms
+  if (predict == "terms") {
+    type_arg <- "terms"
+    scale_arg <- "terms"
+    transform <- FALSE
+
+    # linear models are always on response scale (there is no other)
+  } else if (info$is_linear) {
```

Comment:
Should warn or error if `!predict %in% c("link", "response")`, and not just silently switch.

Reply:
I think this already happens before when we check the input in lines 76 to 91

Reply:
``` r
x <- rnorm(15)
y <- x + rnorm(15)
insight::get_predicted(lm(y ~ x), predict = "count")
#> Warning: `predict` = "count"` is not officially supported by `get_predicted()`.
#>   `predict` will be passed directly to the `predict()` method for the model and
#>   not validated.
#>   Please check the validity and scale of the results.
#>   Set `verbose = FALSE` to silence this warning, or use one of the supported
#>   values for the `predict` argument:
#>     "expectation", "link", "prediction", "classification"
#> Predicted values:
#> 
#>  [1] -1.83681113  1.19301495  0.65079090  2.30813170 -0.08019471  0.17493910
#>  [7] -1.40530023  0.16909396 -0.52268640 -1.34620601 -0.01982088  0.02949426
#> [13]  0.10453808  0.64631889  0.80421114
#> 
#> NOTE: Confidence intervals, if available, are stored as attributes and can be accessed using `as.data.frame()` on this output.
```

<sup>Created on 2022-02-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>